### PR TITLE
0.8 xstyle

### PR DESCRIPTION
**Pull request by sorvell**
* Improvements to cross-scope styling.
   - Support for rules using calc, e.g. calc(var(--a) + var(--b))
   - performance improvments
* support for @extends (https://tabatkins.github.io/specs/css-extend-rule/)

### CODE REVIEW

File: src/lib/style-defaults.html [hunk 1]
```diff
@@ -10,42 +10,45 @@
...
       }
-      return properties;
-    }
 
-    applyCss('');
+    };
+
+    api._styles = [api.style];
+
+    api.applyCss('');
```
Comment (kevinpschaaf) [inline src/lib/style-defaults.html]: Remove vestigial thing.

File: src/lib/style-extends.html [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        this._findExtendor(extend, rule.parent);
+    },
+
+    _extendRule: function(target, source) {
+      if (target.parent !== source.parent) {
+        this._cloneAndAddRuleToParent(source, target.parent);
+      }
+      target.extends = target.extends || (target.extends = []);
+      target.extends.push(source);
+      source.selector = source.selector.replace(this.rx.STRIP, '');
```
Comment (kevinpschaaf) [inline src/lib/style-extends.html]: TODO don't strip to end, there may be `,`'s.

File: src/lib/style-properties.html [hunk 3]
```diff
@@ -121,63 +150,150 @@
...
+          output = this.valueForProperties(rule.propertyInfo.cssText, props);
+        }
+        rule.cssText = output;
+      },
+
+      // Test if the rules in these styles matche the given `element` and if so,
+      // collect any custom properties into `props`.
+      propertiesFromStyles: function(styles, element) {
+        var props = {}, self = this;
+        styleUtil.forRulesInStyles(styles, function(rule) {
```
Comment (kevinpschaaf) [inline src/lib/style-properties.html]: TODO: the rules list itself could be shorter (pre-filter on rules with properties)

File: src/lib/style-properties.html [hunk 4]
```diff
@@ -121,63 +150,150 @@
...
+            // apply css after the scope style of the element to help with
+            // style predence rules.
+            style = styleUtil.applyCss(cssText, selector, 
+              nativeShadow ? element.root : null, element._scopeStyle);
+          }
+        }
+        // ensure this style is our custom style and increment its use count.
+        if (style) {
+          style._useCount = style._useCount || 0;
+          style._useCount++;
```
Comment (kevinpschaaf) [inline src/lib/style-properties.html]: Test if updateStyles with no changes increments useCount for same style element... should be no-op.

File: src/standard/styling.html [hunk 5]
```diff
@@ -11,55 +11,64 @@
...
           }
         }
-        cssText += this._cssFromModule(this.is);
-        var styles = [];
+        cssText += styleUtil.cssFromModule(this.is);
         if (cssText) {
+          // extends!!
+          if (styleExtends.hasExtends(cssText)) {
+            cssText = styleExtends.transform(cssText);
+          }
```
Comment (kevinpschaaf) [inline src/standard/styling.html]: TODO: don't throw away ast

File: src/standard/x-styling.html [hunk 6]
```diff
@@ -10,196 +10,118 @@
...
+        // is available.
+        var notCached = !style;
+        style = this._applyStyleProperties(style);
+        // no cache so store in cache
+        if (notCached) {
+          scope._styleCache.store(this.is, style, this._scopeSelector,
+            scopeProps, this._styles);
+          styleCache.store(this.is, style, this._scopeSelector,
+           this._styleProperties, this._styles);
+        }
```
Comment (kevinpschaaf) [inline src/standard/x-styling.html]: TODO: global cache key should only be the used props

File: src/standard/x-styling.html [hunk 7]
```diff
@@ -10,196 +10,118 @@
...
+        var props = Object.create(scope._styleProperties);
+        // mixin properties matching this element in scope
+        scopeProps = scopeProps || 
+          propertyUtils.propertiesFromStyles(scope._styles, this);
+        this.mixin(props, scopeProps);
+        // finally mixin properties inherent to this element
+        this.mixin(props, propertyUtils.rootPropertiesFromStyles(this._styles));
+        // reify properties
+        propertyUtils.reify(props);
+        this._styleProperties = props;
```
Comment (kevinpschaaf) [inline src/standard/x-styling.html]: Reify only own properties of props (i.e. scopeProps & rootProps)

**Comment by tjsavage** [comment]
Closing this here - we'll have to rebase this on the latest master and merge into the `master` branch

**Pull request closed without merging**